segmentation/utils/spatial_analysis.py

```diff
@@ -107,3 +111,79 @@ def calculate_channel_spatial_enrichment(dist_matrix, marker_thresholds, all_pat
         # stats.append((z, muhat, sigmahat, p, h, adj_p, marker_titles))
         stats.loc[point, :, :] = stats_xr.values
     return values, stats
+
+
+def calculate_phenotype_spatial_enrichment(all_patient_data, dist_mat, points=None,
+                                           bootstrap_num=1000, dist_lim=100):
+    """Spatial enrichment analysis based on cell phenotypes to find significant interactions between different
+    cell types, looking for both positive and negative enrichment. Uses bootstrapping to permute cell labels randomly.
+
+    Args:
+        all_patient_data: data including points, cell labels, and
+            cell expression matrix for all markers
+        dist_mat: cells x cells matrix with the euclidian
+            distance between centers of corresponding cells
+        points: patient labels to include in analysis. If argument is none, default is all labels used
+        bootstrap_num: number of permutations for bootstrap. Default is 1000
+        dist_lim: cell proximity threshold. Default is 100
+
+    Returns:
+        values: a list with each element consisting of a tuple of
+            closenum and closenumrand for each point included in the analysis
+        stats: an Xarray with dimensions (points, stats, number of markers, number of markers) The included stats
+            variables are:
+            z, muhat, sigmahat, p, h, adj_p, and marker_titles for each point in the analysis"""
+
+    patient_idx = "SampleID"
+    cell_type_idx = "cell_type"
+    flowsom_idx = "FlowSOM_ID"
+
+    # Setup input and parameters
+    if points is None:
+        points = list(set(all_patient_data[patient_idx]))
+        num_points = len(points)
+    else:
+        num_points = len(points)
+
+    values = []
+
+    # Error Checking
+    if not np.isin(points, all_patient_data[patient_idx]).all():
+        raise ValueError("Points were not found in Expression Matrix")
+
+    pheno = all_patient_data[[cell_type_idx, flowsom_idx]].drop_duplicates()
```

Comment:
The point of deleting pheno and merging it into the master CSV is so that it removes an additional variable. It defeats the purpose if we then index into the master CSV to recreate pheno. If you need to extract the phenotype names for plotting, you can do that directly by subsetting the master CSV. That way, we avoid the need of having pheno at all.